chainercv/utils/bbox/non_maximum_suppression.py

```diff
@@ -0,0 +1,202 @@
+        assert isinstance(options, tuple)
+        kernel_code = cupy.cuda.compile_with_cache(code, options=options)
+        return kernel_code.get_function(kernel_name)
+
+
+def non_maximum_suppression(bbox, thresh, score=None,
+                            limit=None):
+    """Suppress bounding boxes according to their Jaccard overlap.
+
+    This method checks each bounding box sequentially and selects the bounding
+    box if the Jaccard overlap between the bounding box and the previously
+    selected bounding boxes is less than :obj:`thresh`. This method is
+    mainly used as postprocessing of object detection.
+    The bounding boxes are selected from ones with higher scores.
+    If :obj:`score` is not provided as an argument, the bounding box
+    is ordered by its index in ascending order.
+
+    The bounding boxes are expected to be packed into a two dimensional
+    tensor of shape :math:`(R, 4)`, where :math:`R` is the number of
+    bounding boxes in the image. The second axis represents attributes of
+    the bounding box. They are :obj:`(x_min, y_min, x_max, y_max)`,
+    where the four attributes are coordinates of the bottom left and the
+    top right vertices.
+
+    :obj:`score` is a float array of shape :math:`(R,)`. Each score indicates
+    confidence of prediction.
+
+    This function accepts both :obj:`numpy.ndarray` and :obj:`cupy.ndarray` as
+    inputs. Please note that both :obj:`bbox` and :obj:`score` need to be
+    same type.
+    The output is same type as the type of the inputs.
+
+    Args:
+        bbox (array): Bounding boxes to be transformed. The shape is
+            :math:`(R, 4)`. :math:`R` is the number of bounding boxes.
+        thresh (float): Thresold of Jaccard overlap.
+        score (array): An array of confidences whose shape is :math:`(R,)`.
+        limit (int): The upper bound of the number of the output bounding
+            boxes. If it is not specified, this method selects as many
+            bounding boxes as possible.
+
+    Returns:
+        array:
+        An array with indices of bounding boxes that are selected. \
+        The shape of this array is :math:`(R,)` and its dtype is\
+        :obj:`numpy.int32`.
+
+    """
+
+    xp = cuda.get_array_module(bbox)
+    if xp == np:
+        return _non_maximum_suppression_cpu(bbox, thresh, score, limit)
+    else:
+        return _non_maximum_suppression_gpu(bbox, thresh, score, limit)
+
+
+def _non_maximum_suppression_cpu(bbox, thresh, score=None, limit=None):
+    if len(bbox) == 0:
+        return np.zeros((0,), dtype=np.int32)
+
+    bbox_area = np.prod(bbox[:, 2:] - bbox[:, :2], axis=1)
+    if score is not None:
+        order = score.argsort()[::-1]
+        bbox = bbox[order]
+
+    selec = np.zeros(bbox.shape[0], dtype=bool)
+    for i, b in enumerate(bbox):
+        lt = np.maximum(b[:2], bbox[selec, :2])
+        rb = np.minimum(b[2:], bbox[selec, 2:])
+        area = np.prod(rb - lt, axis=1) * (lt < rb).all(axis=1)
```

Comment:
Since `selec.nonzero()` returns indices that are not in order with `score.argsort()`, it can not be written as simple as you wrote.
I think reordering `bbox` before for loop is easier to read.